### PR TITLE
Some API clarifications

### PR DESCRIPTION
**Pull request by pdillinger**
Summary:
* Clarify that RocksDB is not exception safe on many of our callback
and extension interfaces
* Clarify FSRandomAccessFile::MultiRead implementations must accept
non-sorted inputs (see #8953)
* Clarify ConcurrentTaskLimiter and SstFileManager are not (currently)
extensible interfaces
* Mark WriteBufferManager as `final`, so it is then clearly not a
callback interface, even though it smells like one
* Clarify TablePropertiesCollector Status returns are mostly ignored

Test Plan: comments only (except WriteBufferManager final)

### CODE REVIEW

File: include/rocksdb/concurrent_task_limiter.h [hunk 1]
```diff
@@ -17,6 +17,8 @@
 
 namespace ROCKSDB_NAMESPACE {
 
+// This is NOT a callback interface but a public interface for result of
+// NewConcurrentTaskLimiter. Any derived classes must be RocksDB internal.
```
[the remark this answers is not in the document]
Comment (ajkr) [reply include/rocksdb/concurrent_task_limiter.h:21]: > OK, we can fix this my putting GetToken in the public API.

What function might be implemented differently from ConcurrentTaskLimiterImpl?

**Review by ajkr — APPROVED** [review]
LGTM

File: include/rocksdb/compaction_filter.h [hunk 2]
```diff
@@ -24,7 +24,10 @@ class SliceTransform;
 
 // CompactionFilter allows an application to modify/delete a key-value during
 // table file creation.
-
+//
+// RocksDB callbacks are NOT exception-safe. A callback completing with an
+// exception can lead to undefined behavior in RocksDB, including data loss,
+// unreported corruption, deadlocks, and more.
```
[the remark this answers is not in the document]
Comment (pdillinger) [reply include/rocksdb/compaction_filter.h:30]: Let's suppose two different people involved in application X built on RocksDB do the following: Person 1 handles the build integration and Person 2 writes source code using and extending RocksDB. What are the "other things" Person 2 needs to know about that specifically relate to RocksDB? I can't think of how any you have listed apply to Person 2 specifically in their interaction with RocksDB. Limitations due to compiler options, RTTI availability, etc. will apply project-wide. Your suggestions seem irrelevant at the source code API level.

File: include/rocksdb/concurrent_task_limiter.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (pdillinger) [reply include/rocksdb/concurrent_task_limiter.h:21]: I don't really care whether it's extensible or not. We could idiomatically enforce "internal extension only" with private ctor and friend the Impl class.

In either case, I could envision someone wanting public access to GetToken, just as people use our RateLimiter to rate limit other things along with RocksDB.

File: include/rocksdb/write_buffer_manager.h [hunk 3]
```diff
@@ -34,7 +34,7 @@ class StallInterface {
   virtual void Signal() = 0;
 };
 
-class WriteBufferManager {
+class WriteBufferManager final {
```
Comment (hx235) [inline include/rocksdb/write_buffer_manager.h:37]: I like this! One of my previous PR had this assumption in mind that WBM should not be extended and made a API implement change in [here](https://github.com/facebook/rocksdb/pull/8506/files#diff-5531a9905804e3a40070a32c4f36e3952d09b23c6ca42099074fc180b02d7954R77). It's good to explicitly mark it final just to confirm my assumption!

File: include/rocksdb/concurrent_task_limiter.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (pdillinger) [reply include/rocksdb/concurrent_task_limiter.h:21]: Non-trivial issue because of the Token class. Postponing to another PR.

**Pull request closed without merging**